codegen/render/funcs.go

```diff
@@ -562,6 +571,341 @@ func fromJSON(str string) map[string]interface{} {
+func makePointerFieldsNullable(
+	t reflect.Type,
+	schema *jsonschema.Schema,
+	customTypeMapper customTypeMapper,
+) {
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		jsonPropertyName := field.Name
+
+		// check if the type has been renmamed via the json tag
+		if jsonTag, ok := field.Tag.Lookup("json"); ok {
+			jsonTagSegments := strings.Split(jsonTag, ",")
+			jsonPropertyName = jsonTagSegments[0]
+		}
+
+		if field.Type.Kind() == reflect.Pointer {
+			f, ok := schema.Properties.Get(jsonPropertyName)
+			if !ok {
+				continue
+			}
+			fieldSchema := f.(*jsonschema.Schema)
+
+			// if there's a custom type mapping for the field, delegate to that
+			// whether it wants the field to be nullable
+			if customTypeMapper(field.Type.Elem(), fieldSchema) != nil {
+				continue
+			}
+
+			schema.Properties.Set(jsonPropertyName, wrapAsNullable(fieldSchema))
+		}
+	}
+}
+
+// mergeJsonSchemaProperties Adds the properties from the source json schema
+// to the target
+func mergeJsonSchemaProperties(
+	target *jsonschema.Schema,
+	src *jsonschema.Schema,
+) {
+	if target.Properties == nil {
+		target.Properties = orderedmap.New()
+	}
+
+	for _, prop := range src.Properties.Keys() {
+		val, _ := src.Properties.Get(prop)
+		target.Properties.Set(prop, val)
+	}
+}
+
+// jsonSchemaForOperator creates the json schema for the operator's values.
+//
+// The resulting schema will have the values nested at the appropriate path,
+// taking into account the operator's name and value path. This means it can be
+// added directly to the base object's schema
+func jsonSchemaForOperator(
+	reflector *jsonschema.Reflector,
+	o *values.UserOperatorValues,
+) *jsonschema.Schema {
+	schema := reflector.Reflect(o.Values)
+	schema.Version = "" // will be set on parent object
+
+	// if the opeartor defines custom values, add the properties to the schema
+	if o.CustomValues != nil {
+		mergeJsonSchemaProperties(schema, reflector.Reflect(o.CustomValues))
+	}
+
+	// nest the operator values schema in the correct place
+	path := []string{strcase.ToLowerCamel(o.Name)}
+	if o.ValuePath != "" {
+		splitPath := strings.Split(o.ValuePath, ".")
+		reverse(splitPath)
+		path = append(splitPath, path...)
+	}
+	schema = nestSchemaAtPath(schema, path...)
+
+	return schema
+}
+
+// returns a new schema, nested within an object at the given path
+func nestSchemaAtPath(schema *jsonschema.Schema, path ...string) *jsonschema.Schema {
+	for _, p := range path {
+		parentSchema := new(jsonschema.Schema)
+		parentSchema.Type = "object"
+		parentSchema.Properties = orderedmap.New()
+		parentSchema.Properties.Set(p, schema)
+		schema = parentSchema
+	}
+
+	return schema
+}
+
+func indentJson(src string) string {
+	var buf bytes.Buffer
+	err := json.Indent(&buf, []byte(src), "", "  ")
+	if err != nil {
+		panic(err)
+	}
+	return buf.String()
+}
+
+// reverse mutates a slice of strings to reverse the order.
+func reverse(s []string) {
```

Comment:
one would think golang standard lib would include this....not that I could find it.

Reply:
yeah I googled for quite a while and I couldn't find something that does it :(

Reply:
`Implementation is left to the reader.`

Reply:
wow